seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumRun.scala

```diff
@@ -0,0 +1,171 @@
+        sendAll()
+      }
+    }
+
+    private final def fetchAndProcess(): Unit = {
+      val data = driver
+        .executeScript(s"return $intf.fetch();")
+        .asInstanceOf[java.util.Map[String, java.util.List[String]]]
+        .asScala
+
+      data("consoleLog").asScala.foreach(streams.out.println _)
+      data("consoleError").asScala.foreach(streams.err.println _)
+
+      onMessage.foreach(cb => data("msgs").asScala.foreach(cb))
+
+      val errs = data("errors").asScala
+      if (errs.nonEmpty) throw new SeleniumRun.WindowOnErrorException(errs.toList)
+    }
+
+    private final def isInterfaceUp() =
+      driver.executeScript(s"return !!$intf;").asInstanceOf[Boolean]
+  }
+
+  handler.start()
+
+  def future: Future[Unit] = promise.future
+
+  def close(): Unit = wantClose = true
+
+  def send(msg: String): Unit = {
+    assert(onMessage.isDefined)
+    sendQueue.offer(msg)
+  }
+}
+
+private[selenium] object SeleniumRun {
+  private lazy val validator = {
+    RunConfig.Validator()
+      .supportsInheritIO()
+      .supportsOnOutputStream()
+  }
+
+  def apply(newDriver: () => WebDriver with JavascriptExecutor, input: Input,
+      config: SeleniumJSEnv.Config, runConfig: RunConfig,
+      onMessage: Option[String => Unit]): JSComRun = {
+    validator.validate(runConfig)
+
+    val scripts = input match {
+      case Input.ScriptsToLoad(s) => s
+      case _                      => throw new UnsupportedInputException(input)
+    }
+
+    try {
+      withCleanup(FileMaterializer(config.materialization))(_.close()) { m =>
+        val setupURL = m.materialize("setup.js",
+          JSSetup.setupCode(enableCom = onMessage.isDefined))
+
+        val allScriptURLs = setupURL :: scripts.map(m.materialize)
+
+        val page = m.materialize("scalajsRun.html", htmlPage(allScriptURLs))
+
+        withCleanup(newDriver())(maybeCleanupDriver(_, config)) { driver =>
+          driver.navigate().to(page)
+
+          withCleanup(OutputStreams.prepare(runConfig))(_.close()) { streams =>
+            new SeleniumRun(driver, config, streams, m, onMessage)
+          }
+        }
+      }
+    } catch {
+      case NonFatal(t) =>
+        JSComRun.failed(t)
+
+      case t: NotImplementedError =>
+        /* In Scala 2.10.x, NotImplementedError was considered fatal.
+         * We need this case for the conformance tests to pass on 2.10.
+         */
+        JSComRun.failed(t)
+    }
+  }
+
+  private def withCleanup[V, R](mk: => V)(cleanup: V => Unit)(body: V => R): R = {
+    val v = mk
+    try {
+      body(v)
+    } catch {
+      case t: Throwable =>
+        cleanup(v)
+        throw t
+    }
+  }
+
+  private def maybeCleanupDriver(d: WebDriver, config: SeleniumJSEnv.Config) =
+    if (!config.keepAlive) d.close()
+
+  private def htmlPage(scripts: Seq[java.net.URL]): String = {
+    val scriptTags =
+      scripts.map(path => s"<script src='${path.toString}'></script>")
```

Comment:
`path.toASCIIString`?

Reply:
URL does not have a `toASCIIString` method. Did you mean `toURI.toASCIIString`?

Reply:
Ah right. No leave it as is, then. `URL.toString` is probably correct. (it also seems like this was inherited from the previous version anyway)